src/cascadia/WindowsTerminal/AppHost.cpp

```diff
@@ -126,70 +126,115 @@ void AppHost::LastTabClosed(const winrt::Windows::Foundation::IInspectable& /*se
+        // optional parameter so give two UINTs.
+        UINT dpix = USER_DEFAULT_SCREEN_DPI;
+        UINT dpiy = USER_DEFAULT_SCREEN_DPI;
+        // If this fails, we'll use the default of 96.
+        GetDpiForMonitor(hmon, MDT_EFFECTIVE_DPI, &dpix, &dpiy);
+
+        // We need to check if the top left point of the titlebar of the window is winthin any screen
+        RECT offScreenTestRect;
+        offScreenTestRect.left = proposedRect.left;
+        offScreenTestRect.top = proposedRect.top;
+        offScreenTestRect.right = offScreenTestRect.left + 1;
+        offScreenTestRect.bottom = offScreenTestRect.top + 1;
+
+        bool isTitlebarIntersectWithMonitors = false;
+        EnumDisplayMonitors(
+            nullptr, &offScreenTestRect, [](HMONITOR, HDC, LPRECT, LPARAM lParam) -> BOOL {
+                auto intersectWithMonitor = reinterpret_cast<bool*>(lParam);
+                *intersectWithMonitor = true;
+                // Continue the enumeration
+                return TRUE;
+            },
+            reinterpret_cast<LPARAM>(&isTitlebarIntersectWithMonitors));
+
+        if (!isTitlebarIntersectWithMonitors)
         {
-            // If we failed to get the correct window size for whatever reason, log
-            // the error and go on. We'll use whatever the control proposed as the
-            // size of our window, which will be at least close.
-            LOG_LAST_ERROR();
-            nonClient = Viewport::FromDimensions({ _currentWidth,
-                                                   _currentHeight })
-                            .ToRect();
+            // If the title bar is out-of-screen, we set the initial position to
+            // the top left corner of the nearest monitor
+            proposedRect.left = monitorInfo.rcWork.left;
+            proposedRect.top = monitorInfo.rcWork.top;
         }
-    }
 
-    const auto adjustedHeight = nonClient.bottom - nonClient.top;
-    const auto adjustedWidth = nonClient.right - nonClient.left;
+        auto initialSize = _app.GetLaunchDimensions(dpix);
+
+        const short _currentWidth = Utils::ClampToShortMax(
+            static_cast<long>(ceil(initialSize.X)), 1);
+        const short _currentHeight = Utils::ClampToShortMax(
+            static_cast<long>(ceil(initialSize.Y)), 1);
+
+        // Create a RECT from our requested client size
+        auto nonClient = Viewport::FromDimensions({ _currentWidth,
+                                                    _currentHeight })
+                             .ToRect();
+
+        // Get the size of a window we'd need to host that client rect. This will
+        // add the titlebar space.
+        if (_useNonClientArea)
+        {
+            // If we're in NC tabs mode, do the math ourselves. Get the margins
+            // we're using for the window - this will include the size of the
+            // titlebar content.
+            auto pNcWindow = static_cast<NonClientIslandWindow*>(_window.get());

```

Comment:
I think this can be const.

```suggestion
            const auto pNcWindow = static_cast<NonClientIslandWindow*>(_window.get());
```

Reply:
changed